### PR TITLE
compile_statement restructure (remake of #293)

### PR DESCRIPTION
**Pull request by CAS-ual-TY**
Remake of #293

The following needs to be done in my opinion for this to be merge-able:
- [ ] Compiler class should be done to prevent heavy merge conflicts for students
- [x] #314 needs to be fixed (`make 32-bit` fails right now on the code changes)
- [x] Obviously it needs to be approved

### Changes

This is a remake of the "big pr" that has been sitting idle for almost a year. However, there are no code disadvantages now.

Very important: This can not be merged during a semester. Otherwise there will be conflicts for anyone who has implemented/is working on any assignment which will induce some heavy headaches.

The changes are the following:

- Split current compile_statement into compile_statement (while, if, return, compile_assignment) and compile_assignment (call, assignments to variables)
- Restructure compile_assignment code so that assignments to `uint64_t` and `uint64_t*` (pointers) share the same control flow with pointers simply being dereferenced additionally
- This new structure is also more similar to compile_factor which is already using a more streamlined approach shared for pointers and non-pointers
- The new code is a lot shorter, easier to understand and has less if-else-cascades-depth

Additionally, address/value loading of variables was properly structured into methods and also there is no more sharing between `VARIABLE` and `BIGINT` (eg. `load_variable_or_big_int(...)`):
- `load_variable_or_big_int` split into `load_variable` and `load_big_int`
- `get_variable_or_big_int` was changed to `get_variable` (`get_big_int` not necessary as it would only be a single statement (call) which would also only be referenced a single time, so I just did said statement directly on the reference)
- New method: `load_variable_address` to load an address into a temporary, instead of the value at said address
- Any `if (is_signed_integer(offset, 12)) { ...` constructs are now only in said 3 methods (`load_variable`, `load_big_int`, `load_variable_address`)
- The result is clear responsibilities for methods and said methods being more versatile and usable, especially for the array and struct assignments

Impact for students (my personal opinion):
- I think doing the array assignments is a lot easier now; The relevant code is so much more clear now and these methods are useful and easy to understand
- However, right now the arrays are basically only `uint64_t[]` arrays. With this new addition, `uint64_t*[]` arrays would now also be very easy to do and barely require any additional code (just type checks, basically).

### CODE REVIEW

File: selfie.c [hunk 1]
```diff
@@ -693,10 +693,12 @@ void syntax_error_unexpected();
...
 
-void      load_small_and_medium_integer(uint64_t reg, uint64_t value);
-uint64_t* get_variable_or_big_int(char* variable, uint64_t class);
 void      load_upper_base_address(uint64_t* entry);
-uint64_t  load_variable_or_big_int(char* variable, uint64_t class);
+uint64_t* get_variable(char* variable);
+uint64_t  load_variable_address(char* variable);
+uint64_t  load_variable(char* variable);
+void      load_small_and_medium_integer(uint64_t reg, uint64_t value);
+uint64_t  load_big_int(char* variable);
```
Comment (ckirsch) [inline selfie.c:701]: `load_big_integer` please

**Review by ckirsch — COMMENTED** [review]
@CAS-ual-TY what do you think of the following streamlined grammar, inspired by your changes:

```
cstar      = { variable [ "=" [ cast ] [ "-" ] value ] ";" |
               ( "void" | type ) identifier procedure } .

variable   = type identifier .

type       = "uint64_t" [ "*" ] .

cast       = "(" type ")" .

value      = integer_literal | character_literal .

statement  = assignment ";" | if | while | call ";" | return ";" .

assignment = ( [ "*" ] identifier | "*" "(" expression ")" ) "=" expression .

expression = arithmetic [ ( "==" | "!=" | "<" | ">" | "<=" | ">=" ) arithmetic ] .

arithmetic = term { ( "+" | "-" ) term } .

term       = factor { ( "*" | "/" | "%" ) factor } .

factor     = [ cast ] [ "-" ] [ "*" ] ( literal | identifier | call | "(" expression ")" ) .

literal    = value | string_literal .

if         = "if" "(" expression ")"
               ( statement | "{" { statement } "}" )
             [ "else"
               ( statement | "{" { statement } "}" ) ] .

while      = "while" "(" expression ")"
               ( statement | "{" { statement } "}" ) .

procedure  = "(" [ variable { "," variable } [ "," "..." ] ] ")" ( ";" |
             "{" { variable ";" } { statement } "}" ) .

call       = identifier "(" [ expression { "," expression } ] ")" .

return     = "return" [ expression ] .
```

**Comment by ckirsch** [comment]
@CAS-ual-TY do you think there is a near one-to-one mapping of the polished grammar with parser code? That would be awesome.

File: selfie.c [hunk 2]
```diff
@@ -1089,7 +1092,7 @@ uint64_t ELF_HEADER_SIZE = 4096;
 uint64_t ELFCLASS64 = 2;
 uint64_t ELFCLASS32 = 1;
 
-uint64_t MAX_CODE_SIZE = 262144; // 256KB
+uint64_t MAX_CODE_SIZE = 524288; // 256KB
```
Comment (ckirsch) [inline selfie.c:1095]: // 512KB

**Comment by CAS-ual-TY** [comment]
> @CAS-ual-TY do you think there is a near one-to-one mapping of the polished grammar with parser code? That would be awesome.

Yes. I believe I can make that work.

**Comment by ckirsch** [comment]
> > @CAS-ual-TY do you think there is a near one-to-one mapping of the polished grammar with parser code? That would be awesome.
> 
> Yes. I believe I can make that work.

Great! Please have the compile_ procedures appear in the same order as the grammar rules.

Meanwhile I am trying to find that 32-bit bug.

**Comment by CAS-ual-TY** [comment]
Working on it. A few things:

- Not sure if I can properly add `compile_value`, since those are already in get_symbol() which is basically the lowest level
- Suggestion: `arithmetic` -> `sum` and `term` -> `product`

**Comment by ckirsch** [comment]
> Working on it. A few things:
> 
> * Not sure if I can properly add `compile_value`, since those are already in get_symbol() which is basically the lowest level
> * Suggestion: `arithmetic` -> `sum` and `term` -> `product`

Feel free to massage the grammar as well to make everything fit as nicely as possible.

Just don't change the term `term`. It's everywhere in my book...

I am not sure about `arithmetic` but keep it for now.

**Comment by ckirsch** [comment]
@CAS-ual-TY can you pinpoint the commit where 32-bit started failing?

**Comment by CAS-ual-TY** [comment]
> @CAS-ual-TY can you pinpoint the commit where 32-bit started failing?

https://github.com/CAS-ual-TY/selfie/commit/ce2a4edd04e44e750eb6d7d4804300f988cb3542

Not changes to selfie whatsoever.

**Comment by ckirsch** [comment]
@CAS-ual-TY just pushed (hopefully) a fix of the 32-bit issue, waiting for CI.

**Comment by CAS-ual-TY** [comment]
What do you think of this:

```
cstar          = { variable [ initialization ] ";" |
                   ( "void" | type ) identifier procedure } .

initialization = "=" [ cast ] [ "-" ] value .
```

**Comment by ckirsch** [comment]
> @ckirsch What do you think of this:
> 
> ```
> cstar          = { variable [ initialization ] ";" |
>                    ( "void" | type ) identifier procedure } .
> 
> initialization = "=" [ cast ] [ "-" ] value .
> ```
> 
> It would be more closely fitting to the 1-1 grammar mapping

Good!

**Comment by CAS-ual-TY** [comment]
Moved everything into `compiler_restructure` branch.

**Comment by ckirsch** [comment]
@CAS-ual-TY can you try to sync, in grammar and source code, the order in which production rules and parser procedures occur? Any auxiliary procedures should appear where they are needed first.

**Pull request closed without merging**